File: lru-cache/lru-cache.swift
/**
get -> O(1)
put -> O(1)
hash map doubly linked list

1    1
2    3
3    4
4    2


policy: when cache is full -> remove LRU
when you look at a new item, then that item if its in the hash map then put to the top
*/
class ListNode: Hashable {
    var key: Int
    var val: Int
    weak var prev: ListNode?
    var next: ListNode?
    
    init(_ key: Int, _ val: Int) {
        self.key = key
        self.val = val
    }
    
    public static func == (lhs: ListNode, rhs: ListNode) -> Bool {
        return lhs === rhs
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(val)
        hasher.combine(prev?.val ?? 0)
        hasher.combine(next?.val ?? 0)
    } 
}

class LRUCache {
    
    var dict: [Int: ListNode] = [:]
    let capacity: Int
    var head = ListNode(-1, -1)
    var tail = ListNode(-1, -1)

    init(_ capacity: Int) {
        guard capacity > 0 else {
            self.capacity = 1
            return
        }
        self.capacity = capacity
        head.next = tail
        tail.prev = head
    }
    
    // 1<->2<->3<->4... take 3, connect it to current head, set its prev to null, then at your current spot, bind the left's next to the right and the right's prev to the left
    /**
  x<-1<->2<->3<->4->x
    
    x<3>x
    */
    func get(_ key: Int) -> Int {
        if let mid = dict[key] {
            // set left's and right's
            let l = mid.prev
            let r = mid.next
            l?.next = r
            r?.prev = l
            
            // add floating node to beginning of list
            mid.prev = head
            mid.next = head.next
            head.next?.prev = mid
            head.next = mid
            
            return mid.val
        }
        return -1
    }
    
    /**

    x 1 x
    */
    func put(_ key: Int, _ value: Int) {
        if let node = dict[key] {
            node.val = value
            let l = node.prev
            let r = node.next
            l?.next = r
            r?.prev = l
            
            node.prev = head
            node.next = head.next
            head.next?.prev = node
            head.next = node
        } else {
            if dict.count == capacity {
                dict.removeValue(forKey: tail.prev!.key)
                let l = tail.prev?.prev
                let r = tail
                l?.next = r
                r.prev = l
            }
            let newNode = ListNode(key, value)
            dict[key] = newNode

            newNode.prev = head
            newNode.next = head.next
            head.next?.prev = newNode
            head.next = newNode
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * let obj = LRUCache(capacity)
 * let ret_1: Int = obj.get(key)
 * obj.put(key, value)
 */

File: lru-cache/lru-cache.kt
/**
capacity

- doubly linked list
- hashmap
key : key of item
value : node -> val

access -> O(1)
retrieve -> O(1)

head - 6 - 1 - 2 - 4 - 8 - tail

head - tail

get the value:
- bring it to the front -> break the left and right pointers, then add to front

put:
- find node using key
- bring the value to the front
  - if its a new value, add to front
  - if its existing value, bring that to the front
- if youre at capacity, remove that from back


*/
class Node(val key: Int, var v: Int, var next: Node? = null, var prev: Node? = null) {}

class LRUCache(capacity: Int) {
    
    var cache = HashMap<Int, Node>()
    val head = Node(0, 0)
    val tail = Node(0, 0)
    val capacity: Int
    
    init {
        this.capacity = capacity
        head.next = tail
        tail.prev = head
        // h <-> t
    }

    fun get(key: Int): Int {
        val node = cache[key]
        if (node != null) {
            remove(node)
            add(node)
            return node.v
        } else {
            return -1
        }
    }

    fun put(key: Int, value: Int) {
        val node = cache[key]
        if (node != null) {
            node.v = value
            remove(node)
            add(node)
        } else {
            if (cache.size == capacity) {
                val last = tail.prev!!
                remove(last)
                cache.remove(last.key)
            }
            val newNode = Node(key, value)
            add(newNode)
            cache[key] = newNode
        }
    }

    fun remove(m: Node) {
        val l = m.prev
        val r = m.next
        l?.next = r
        r?.prev = l
    }
    
    fun add(m: Node) {
        val l = head
        val r = head.next
        l?.next = m
        m.prev = l
        m.next = r
        r?.prev = m
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
